text/add-bf16-f64f64-and-f80-type.md

```diff
@@ -0,0 +1,121 @@
+impl From<f64> for f64f64 { /* ... */ }
+```
+
+`f64f64` will generate `ppc_fp128` type in LLVM IR.
+
+## `f80` type
+
+`f80` represents the extended precision floating type on x86 targets, with 1 sign bit, 15 bits of exponent and 63 bits of mantissa.
+
+The following `From` traits are implemented in `core::arch::{x86, x86_64}` for conversion between `f64f64` and other floating types:
+
+```rust
+impl From<bf16> for f80 { /* ... */ }
+impl From<f32> for f80 { /* ... */ }
+impl From<f64> for f80 { /* ... */ }
+```
+
+`f80` will generate `x86_fp80` type in LLVM IR.
+
+## `c_longdouble` type in FFI
+
+`core::ffi::c_longdouble` will always represent whatever `long double` does in C. Rust will defer to the compiler backend (LLVM) for what exactly this represents, but it will approximately be:
+
+- 80-bit extended precision (f80) on `x86` and `x86_64`:
+- `f64` double precision with MSVC
+- `f128` quadruple precision on AArch64
+- `f64f64` on PowerPC
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+`bf16` is not a IEEE-754 standard type, so adding it as primitive type may break existing consistency for builtin float types. The truncation after calculation on targets not supporting `bf16` natively also breaks how Rust treats precision loss in other cases.
+
+`c_longdouble` are not uniquely determined by architecture, OS and ABI. On the same target, C compiler options may change what representation `long double` uses.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[half](https://github.com/starkat99/half-rs) crate provides implementation of binary16 and bfloat16 types.
+
+However, besides the disadvantage of usage inconsistency between primitive type and type from crate, there are still issues around those bindings.
+
+The availablity of additional float types depends on CPU/OS/ABI/features of different targets heavily. Evolution of LLVM may also unlock possibility of the types on new targets. Implementing them in compiler handles the stuff at the best location.
+
+Most of such crates defines their type on top of C binding. But extended float type definition in C is complex and confusing. The meaning of `long double`, `_Float128` varies by targets or compiler options. Implementing in Rust compiler helps to maintain a stable codegen interface.
+
+And since third party tools also relies on Rust internal code, implementing additional float types in compiler also help the tools to recognize them.
+
+# Prior art
+[prior-art]: #prior-art
+
+There is a previous proposal on `f16b` type to represent `bfloat16`: https://github.com/rust-lang/rfcs/pull/2690.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+This proposal does not contain information for FFI with C's `_Float128` and `__float128` type. Because they are not so commonly used compared to `long double`, and they are even more complex than the situation of `c_longdouble` (for example, their semantics are different under C or C++ mode).
```

Comment:
```suggestion
This proposal does not contain information for FFI with C's `_Float128` and `__float128` type, because they are not so commonly used compared to `long double`, and they are even more complex than the situation of `c_longdouble` (for example, their semantics are different under C and C++).
```

Reply:
I think this is not the reason, indeed,  on some target `c_longdouble`  needs `_Float128`, The real reason is because we have a different proposal for it. Do not said this as it's misleading

Reply:
PowerPC has option to control what `long double` means: `f64`, `f128` or `f64f64`. But since it is in transition to `f128` as of the time of writing, we can drop a little history burden and set `f128` on little-endian 64-bit targets.

I don't make sure if x86 has similar option. If not, we can confidently introduce `c_longdouble`.